Add documentation of the Application Discovery feature in Control Center (#3540)
vaadin/docs: articles/control-center/application-discovery/index.adoc

---
title: Application Discovery
description: How to use Control Center's Application Discovery feature.
order: 40
---

= Application Discovery

> Control Center is meant to be a central management solution for Vaadin apps deployed in a Kubernetes cluster.

Reviewer: ```suggestion
Control Center is meant to be a central management solution for Vaadin applications deployed in a Kubernetes cluster.
```

If you want to manage a Vaadin application with Control Center, use the Discovery view to add the application to Control Center.
Before an application can be properly managed by Control Center, some preparations need to be done first.

> == Prepare an application for discovery

Reviewer: ```suggestion
== Prepare an Application for Discovery
```

First, you must add the `control-center-starter` dependency to your Vaadin application.
[source,xml]
----
<dependency>
    <groupId>com.vaadin</groupId>
    <artifactId>control-center-starter</artifactId>
    <version>1.0.0</version>
</dependency>
----

> Create a `bootstrap.yaml` file with the application name and kubernetes reload enabled.

Reviewer: ```suggestion
Create a `bootstrap.yaml` file with the application name and Kubernetes reload enabled.
```

[source,yaml]
----
spring:
  application:
    name: control-center-application
  cloud:
    kubernetes:
      reload:
        enabled: true
----

Build a docker image of your application with an appropriate `Dockerfile` file.

[source,shell]
----
docker build -t vaadin/example:0.1.1 .
----

> == Deploy and adopt a Vaadin application

Reviewer: ```suggestion
== Deploy and Adopt a Vaadin Application
```

Once you have a docker image of your application ready, the next step is to deploy your application using Control Center.
In the Application Discovery view, click on "Deploy". This opens the application deployment dialog.

image::images/deploy_button.png[Deploy button]

Fill in the text fields with the data for the application. 

- In the "Application name" field, use the same name as the one in the `bootstrap.yml` file.

> - In the "Image" field, remember to use the correct tag.

Reviewer: Perhaps add an e.g. here with the version number from earlier (0.1.1)?

> - In "Hostname" fill a hostname that can is resolvable by the browser (like the Keycloak hostname).

Reviewer: ```suggestion
- In the "Hostname" field, enter a hostname that can is resolvable by the browser (like the Keycloak hostname).
```

- The "Use a secure connection" checkbox is optional. It determines whether a self-signed certificate is to be created for the application or not.

> NOTE: if the version in the image tag is `latest`, Control Center pulls the image from the public registry.

Reviewer: ```suggestion
NOTE: If the version in the image tag is `latest`, Control Center pulls the image from the public registry.
```

> One you click "Confirm", Control Center creates a deployment using your application's docker image and make it available at the specified hostname as soon as possible.

Reviewer: ```suggestion
One you click "Confirm", Control Center creates a deployment using your application's docker image and makes it available at the specified hostname as soon as possible.
```

image::images/deploy_dialog_1.png[Deploy dialog filled]

You can adopt the application at any moment so it is available for selection in the application selection drop-down. This allows you to select the application for management in other parts of Control Center, like user management.
To adopt an orphaned application, select it and click on "Adopt".

image::images/orphaned.png[Adopt an orphaned application]

When you no longer wish to have Control Center manage an application, select it and click "Delete".

== Scale an Application

> Once an application has been deployed, it is easy to scale it, be it up or down. Select the application to scale and click on "Scale". This opens the scale application dialog. Write down the number of replicas you  wish your cluster to have and click "Confirm". Your application then is automatically scaled to the desired number of replicas.

Reviewer: ```suggestion
Once an application has been deployed, it is easy to scale it up or down. Select the application to scale and click on "Scale". This opens the scale application dialog. Write down the number of replicas you wish your cluster to have, and click "Confirm". Your application then is automatically scaled to the desired number of replicas.
```
